icefoganalytics/travel-authorization
# Travel Authorization: Estimates Calculate on Fly/in Advance

Fixes https://github.com/icefoganalytics/travel-authorization/issues/54

# Context

In order to submit a user should have an estimated cost defined.  The current workflow requires the user to save a draft, navigate to expenses, generate the expenses, and then submit (often the user would need to navigate back to the details to review the advance and select the supervisor).  It would be better to have the estimate generate on the fly and show in the field as part of the approvals.  
![image](https://github.com/icefoganalytics/travel-authorization/assets/89539008/6dca055c-16db-45c9-aba3-7f94ea945d45)

Perhaps put a button beside estimate that will take the user to the estimate tab if they want to change the calculate estimate.  Most will just leave it as is.  This estimate is important when we look to implement the signing authority application.

## Investigation

This is an interesting problem, generally speaking, this would be best solved from a UI perspective, using the "Wizard" pattern.
e.g. several linked pages that the user has to complete, in order.
1. Purpose + Details page
3. Estimate page
4. Supervisor selection and submission page.

A similar thing to a "wizard" using our component system is the ["v-stepper" component](https://v2.vuetifyjs.com/en/components/steppers/); although, wizards have a lot of back-end wiring that goes along with the UI component setup to allow partial completion/save/and resume, etc.

As far as automation goes, I can
1. Put a call-to-action button (or tooltip?) beside the estimate field, that redirects the user to the estimates page.
2. Auto-generate the estimate, if one does not exist, on "save draft" action.

I've been concerned about too much automation here, as the Estimate generation code is too complicated to really trust without having a test suite backing it.

# Implementation

Use some wizard of oz magic to handle users editing estimates, then going back to the Details tab, performs a silent refresh.
Make sure that when a user generates estimates, it performs a blocking refresh, that doesn't reload the page.

# Screenshots

Pre-Estimate Generation
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/570ec6e3-6009-4fb6-8973-b18b6847d2d7)

Estimate Generation Dialog
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/816f7374-9784-4766-b7a3-94df784867e9)

Post-Estimate Generation
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/239ebc46-dcc7-4ee0-a1aa-3f2a1c2b614f)

# Testing Instructions

1. Boot the app via `dev up`
2. Log in to the app at http://localhost:8080
3. Go to the My Travel Requests page via the top nav.
4. Create a new Travel Authorization.
5. Fill in the Details section of the travel authorization.
6. Go to the Approvals section and you will now see a (?) icon by the "Estimated Cost", on hover you get a tooltip about where this information comes from.
7. Note that the "Submit to Supervisor" button is now disabled and has a (?) icon in it that comes with a tooltip about why the button is disabled.
8. Click the new Generate Estimates button beside the "Estimated Cost" field.
9. You will get a dialog to generate the estimate.
10. After you click the button, it blocks until the estimates are generated. It does not break page state, nor reload the whole page. This was very hard to do :P. 
    > Probably hard to test for QA, for me I put a 30 second sleep in the back-end route.
11. You can see now see the "Estimated Cost" is now filled in. 
12. The "Generate Estimates" button has changed to an "Edit Estimate" button that links to the "Estimate" tab.
13. The "Submit to Supervisor" button is now enabled.
14. If you click the "Edit Estimate" button, you will be redirected to the "Estimate" tab.
15. When you edit/create/delete some estimates, then go back to the "Details" tab, the "Estimated Cost" will be up to date, this is the result of some special magic, as seen in the "Implementation" section of this PR.
